README.md
# FFBE-sprite-sheet-assembler
Tool to assemble Final Fantasy Brave Exvius sprite sheets.

Takes the a master sprite atlas png file and uses the information from cvs files
to assemble the sprite sheet.

## Instructions

* Requires [nodejs](https://nodejs.org/en/)

`npm install`

### Usage
```
node build/ffbetool.js num [-a anim] [-c columns] [-e] [-v] [-j] [-g] [-i inDir] [-o outDir]

```
* **num:** (required) the unit ID number, must be the first argument.
* **-a:** The animation name (ie. atk, idle, dead, win, etc.).
* **-c:** The number of columns in the sheet, if not specified the output will be a single-row strip.
* **-e:** If this option is included, the strips will include any empty frames. By default empty frames are excluded.
* **-v:** If this option is included, more information will be printed to console.
* **-j:** This option saves the sheet information in json format saved as a .json file.
* **-g:** If htis option is included, an animated gif will also be outputed.
* **-i:** The input path, defaults to current directory.
* **-o:** The output path, defaults to current directory.


### References
* Code based off of puggsoy's [FFBETool](https://github.com/puggsoy/MiscTools/tree/master/FFBETool/src)
* Original python code for Brave Frontier from [pastebin](http://pastebin.com/vXc0yNRh)
* Information gleaned from: [The VG Resource](https://www.google.com/url?sa=t&rct=j&q=&esrc=s&source=web&cd=9&cad=rja&uact=8&ved=0ahUKEwjU8bHRxsfOAhVL62MKHT6xCLwQFgg5MAg&url=http%3A%2F%2Fwww.vg-resource.com%2Fthread-27841.html&usg=AFQjCNHXVA5Jn78-QtXtJAtpmuZoEAxr_g&sig2=M6vg5hTSpyOJUD2qMuIUsQ&bvm=bv.129759880,d.cGc)



CHANGELOG.md
## [0.3.0]

### New Features

* Added the [-g] option to output an animated gif

## [0.2.3]

### Changes

* Removed the [-n] option
* Bundle the and dependencies into **build/ffbetool.js** so the can be run without requiring `npm install`
* Refactored for maintainability
* Use ES6

## [0.2.2]

### New Features

* Added [-n] option to use the out filename format `$uid_$action`

### Changes

* Default out filename format to `unit_$action_$uid`

## [0.2.1]

### New Features

* Added [-e] option to include empty frames
* Added [-v] option to print out more information to console
* Added [-j] option to save the animation sheet information as json file

### Changes

* Fixed blend mode
* Some optimizations
* Recursively create directories to output path if necessary
* Print less to console by default
* Made the logs to console more informative

## [0.2.0]

### Changes

* Flip part image before applying rotation
* Some optimizations
* Removed the [-e] option to include empty frames
* Removed the [-d] divider option
* Promisification of callbacks

## [0.1.0]

### Notes

* Initial Release


src/Image.js
import Jimp, { intToRGBA, rgbaToInt } from 'jimp';
import { range } from 'lodash';

/**
 * Creates a Promise wrapper around Jimp's constructor.
 *
 * @param {number} width - The image width in pixels
 * @param {number} height - The image height in pixels
 * @return {Promise} - Promise resolving to a Jimp object
 */
export const createImage = function (width, height) {
    return new Promise((resolve, reject) => {
        // eslint-disable-next-line no-new
        new Jimp(width, height, (err, image) => {
            if (err) {
                return reject(err);
            }

            return resolve(image);
        });
    });
};

/**
 * Converts color range of form 0.0-1.0 to 0-255
 *
 * @param {Object} pixel    Object describing the pixel colors in range 0-1
 * @param {number} pixel.r  The red channel
 * @param {number} pixel.g  The green channel
 * @param {number} pixel.b  The blue channel
 * @param {number} pixel.a  The alpha channel
 * @return {Object} The object describing the pixel colors in range 0-255
 */
const convertColorToDecimalRange = function ({ r, g, b, a }) {
    return {
        r: r / 255,
        g: g / 255,
        b: b / 255,
        a: a / 255,
    };
};

/**
 * Converts color range of form 0-255 to 0.0-1.0
 *
 * @param {Object} pixel    Object describing the pixel colors in range 0-255
 * @param {number} pixel.r  The red channel
 * @param {number} pixel.g  The green channel
 * @param {number} pixel.b  The blue channel
 * @param {number} pixel.a  The alpha channel
 * @return {Object} The object describing the pixel colors in range 0-1
 */
const convertColorTo255Range = function ({ r, g, b, a }) {
    return {
        r: Math.round(r * 255),
        g: Math.round(g * 255),
        b: Math.round(b * 255),
        a: Math.round(a * 255),
    };
};

/**
 * Applies a blending of color channels
 *
 * @param {Jimp}    The Jimp image object to apply blending
 * @return {Jimp}   The transformed Jimp image object
 */
export const blend = function (image) {
    const { width, height } = image.bitmap;

    range(width).forEach((col) => {
        range(height).forEach((row) => {
            const { a, r, g, b } =
                convertColorToDecimalRange(intToRGBA(image.getPixelColor(col, row)));

            if (a !== 0) {
                const pixel = convertColorTo255Range({
                    r: r * a,
                    g: g * a,
                    b: b * a,
                    a: (r + g + b) / 3,
                });

                image.setPixelColor(rgbaToInt(pixel.r, pixel.g, pixel.b, pixel.a), col, row);
            }
        });
    });

    return image;
};

// mask : hex
// color : hex
// findColor : boolean

// attempt to implement haxe's
// function getColorBoundsRect( mask : UInt, color : UInt, ?findColor : Bool ) : Rectangle
/*
   Determines a rectangular region that either fully encloses all pixels of a specified color
   within the bitmap image (if the findColor parameter is set to true) or fully encloses all pixels
   that do not include the specified color (if the findColor parameter is set to false).

   For example, if you have a source image and you want to determine the rectangle of the image
   that contains a nonzero alpha channel, pass {mask: 0xFF000000, color: 0x00000000} as parameters.
   If the findColor parameter is set to true, the entire image is searched for the bounds of pixels
   for which (value & mask) == color (where value is the color value of the pixel). If the
   findColor parameter is set to false, the entire image is searched for the bounds of pixels for
   which (value & mask) != color (where value is the color value of the pixel). To determine white
   space around an image, pass {mask: 0xFFFFFFFF, color: 0xFFFFFFFF} to find the bounds of nonwhite
   pixels.
*/
export const getColorBoundsRect = function (image, mask, color, findColor) {
    // findColor : value & mask === color
    // !findColor : value & mask !== color
    const { width, height } = image.bitmap;

    const extremities = range(height).reduce((acc, row) =>
        range(width).reduce((obj, col) => {
            const pixelColor = image.getPixelColor(col, row);

            // eslint-disable-next-line no-bitwise
            if ((findColor && (pixelColor & mask === color)) || (pixelColor & mask !== color)) {
                const { minx, miny, maxx, maxy } = obj;
                return {
                    minx: col < minx ? col : minx,
                    miny: row < miny ? row : miny,
                    maxx: col > maxx ? col : maxx,
                    maxy: row > maxy ? row : maxy,
                };
            }
            return obj;
        }, acc)
    , { minx: width, maxx: 0, miny: height, maxy: 0 });

    const { minx: x, miny: y, maxx, maxy } = extremities;

    return {
        x,
        y,
        width: Math.max(0, (maxx - x) + 1),
        height: Math.max(0, (maxy - y) + 1),
    };
};

export default {
    createImage,
    blend,
    getColorBoundsRect,
};


main.js
var Promise = require('bluebird');
var fs = Promise.promisifyAll(require('fs'));
var mkdirp = Promise.promisifyAll(require('mkdirp'));
var path = require('path');
var _ = require('lodash');
var Jimp = require('jimp');

var usage = 'Usage: main num [-a anim] [-c columns] [-e] [-v] [-j] [-n] [-i inDir] [-o outDir]';

var ffbeTool = function () {
   this.id = -1;
   this.animName = '';
   this.columns = 0;
   this.inputPath = '.';
   this.outputPath = '.';
   this.includeEmpty = false;
   this.verbose = false;
   this.saveJson = false;
   this.customFilename = false;  // uid_action  | default: unit_action_uid

   this.cggPath = null;
   this.pngPath = null;

   this.frames = [];
};

// mask : hex
// color : hex
// findColor : boolean

// attempt to implement haxe's
// function getColorBoundsRect( mask : UInt, color : UInt, ?findColor : Bool ) : Rectangle
/*
   Determines a rectangular region that either fully encloses all pixels of a specified color within the bitmap image (if the findColor parameter is set to true) or fully encloses all pixels that do not include the specified color (if the findColor parameter is set to false).

   For example, if you have a source image and you want to determine the rectangle of the image that contains a nonzero alpha channel, pass {mask: 0xFF000000, color: 0x00000000} as parameters. If the findColor parameter is set to true, the entire image is searched for the bounds of pixels for which (value & mask) == color (where value is the color value of the pixel). If the findColor parameter is set to false, the entire image is searched for the bounds of pixels for which (value & mask) != color (where value is the color value of the pixel). To determine white space around an image, pass {mask: 0xFFFFFFFF, color: 0xFFFFFFFF} to find the bounds of nonwhite pixels.
*/
var getColorBoundsRect = function (image, mask, color, findColor) {
   //findColor : value & mask === color
   //!findColor : value & mask !== color

   var width = image.bitmap.width;
   var height = image.bitmap.height;

   var minx = image.bitmap.width;
   var maxx = 0;
   var miny = image.bitmap.height;
   var maxy = 0;

   _.each(_.range(height), function (y) {
      _.each(_.range(width), function (x) {
         var value = image.getPixelColor(x, y);

         if (findColor) {
            if (value & mask === color) {
               if (x < minx) {
                  minx = x;
               }
               if (x > maxx) {
                  maxx = x;
               }
               if (y < miny) {
                  miny = y;
               }
               if (y > maxy) {
                  maxy = y;
               }
            }
         } else {
            if (value & mask !== color) {
               if (x < minx) {
                  minx = x;
               }
               if (x > maxx) {
                  maxx = x;
               }
               if (y < miny) {
                  miny = y;
               }
               if (y > maxy) {
                  maxy = y;
               }
            }
         }
      }); // end each y
   }); // end each x

   return {
      x: minx,
      y: miny,
      width: Math.max(0, maxx - minx + 1),
      height: Math.max(0, maxy - miny + 1)
   };
};

// Promise wrap Jimp constructor
var createImage = function (width, height) {
   return new Promise(function (resolve, reject) {
      new Jimp(width, height, function (err, image) {
         if (err) {
            return reject(err);
         }

         return resolve(image);
      });
   });
};

var blend = function (image) {
   _.each(_.range(image.bitmap.width), function (x) {
      _.each(_.range(image.bitmap.height), function (y) {
         var pixel = convertColorRange01(Jimp.intToRGBA(image.getPixelColor(x, y)));

         if (pixel.a !== 0) {
            pixel.r = pixel.r * pixel.a;
            pixel.g = pixel.g * pixel.a;
            pixel.b = pixel.b * pixel.a;
            pixel.a = (pixel.r + pixel.g + pixel.b) / 3;

            pixel = convertColorRange255(pixel);
            image.setPixelColor(Jimp.rgbaToInt(pixel.r, pixel.g, pixel.b, pixel.a), x, y);
         }
      }); // end y
   }); // end x

   return image;
};

var convertColorRange01 = function (r, g, b, a) {
   var color = {};
   if (typeof r === 'object') {
      color = r;
   } else {
      color.r = r;
      color.g = g;
      color.b = b;
      color.a = a;
   }

   return {
      r: color.r / 255,
      g: color.g / 255,
      b: color.b / 255,
      a: color.a / 255
   };
};

var convertColorRange255 = function (r, g, b, a) {
   var color = {};
   if (typeof r === 'object') {
      color = r;
   } else {
      color.r = r;
      color.g = g;
      color.b = b;
      color.a = a;
   }

   return {
      r: Math.round(color.r * 255),
      g: Math.round(color.g * 255),
      b: Math.round(color.b * 255),
      a: Math.round(color.a * 255)
   };
};

ffbeTool.prototype = {
   processCommandArgs: function (argv) {
      var i = 3, len = argv.length;

      while (i < len) {
         switch(argv[i]) {
            case '-a':
               this.animName = argv[++i];
               break;
            case '-c':
               this.columns = parseInt(argv[++i]);
               break;
            case '-e':
               this.includeEmpty = true;
               break;
            case '-i':
               this.inputPath = argv[++i];
               break;
            case '-o':
               this.outputPath = argv[++i];
               break;
            case '-v':
               this.verbose = true;
               break;
            case '-j':
               this.saveJson = true;
               break;
            case '-n':
               this.customFilename = true;
               break;
         }
         i += 1;
      }
   },

   readCggAsync: function (unitID) {
      var verbose = this.verbose;
      this.cggPath = path.join(this.inputPath, 'unit_cgg_' + unitID + '.csv');

      console.info('Loading ' + this.cggPath + '...');

      return fs.readFileAsync(this.cggPath, 'utf8')
         .then(function (data) {
            var datasplit = data.replace('\r').split('\n');

            var processDataLine = function (line, index) {
               return new Promise(function (resolve, reject) {
                  var params = line.split(',');
                  params.splice(params.length - 1, 1);

                  var anchor = 0,
                     count = 0,
                     parts = [],
                     part = null,
                     i = 0;

                  if (params.length >= 2) {
                     anchor = parseInt(params[0]);
                     count = parseInt(params[1]);
                     parts = [], i = 2;

                     _.each(_.range(count), function (partInd) {
                        part = {};

                        part.xPos = parseInt(params[i++]);
                        part.yPos = parseInt(params[i++]);
                        part.nextType = parseInt(params[i++]);
                        part.flipX = false;
                        part.flipY = false;

                        switch (part.nextType) {
                           case 0:
                              break;
                           case 1:
                              part.flipX = true;
                              break;
                           case 2:
                              part.flipY = true;
                              break;
                           case 3:
                              part.flipX = true;
                              part.flipY = true;
                              break;
                           default:
                              console.log("Invalid next type!");
                              return reject("Invalid next type!");
                        } // end switch

                        part.blendMode = parseInt(params[i++]);
                        part.opacity = parseInt(params[i++]);
                        part.rotate = parseInt(params[i++]);
                        part.imgX = parseInt(params[i++]);
                        part.imgY = parseInt(params[i++]);
                        part.imgWidth = parseInt(params[i++]);
                        part.imgHeight = parseInt(params[i++]);
                        part.pageID = parseInt(params[i++]);

                        part.lineIndex = index;
                        part.line = line;
                        part.index = partInd;

                        parts.push(part);
                     }); // end inner _.each

                     return resolve(parts.reverse());
                  } else {
                     if (verbose) {
                        console.log('line ' + index + ' : params.length was less than 2');
                     }
                     return resolve(null);
                  }
               }); // end Promise
            }; // end processDataLine

            var processing = datasplit.map(processDataLine);
            var results = Promise.all(processing);

            return results.then(function (frames) {
               return {
                  unitID: unitID,
                  frames: frames
               };
            });
         }); // end readFileAsync
   },

   readPngAsync: function (data) {
      var verbose = this.verbose;
      var frames = data.frames;
      var unitID = data.unitID;

      var inputPath = this.inputPath;

      var pngPath = path.join(this.inputPath, 'unit_anime_' + unitID + '.png');
      var png = Jimp.read(pngPath);
      var cgsPath;

      if (this.animName) {
         cgsPath = path.join(inputPath, 'unit_' + this.animName + '_cgs_' + unitID + '.csv');
         return png.then(_.bind(function (image) {
            this.makeStrip(cgsPath, frames, image);
         }, this));
      } else {
         console.info('* No animName provided *');
         return png.then(_.bind(function (image) {
            fs.readdirAsync(this.inputPath).map(_.bind(function (file) {

               var extension = path.extname(file);
               cgsPath = path.join(this.inputPath, file);

               if (extension === '.csv' && file.indexOf('_cgs_') >= 0 &&
                  file.indexOf(unitID) >= 0) {

                  if (verbose) {
                     console.log(' -- ' + file + ' is target cgs');
                  }
                  this.makeStrip(cgsPath, frames, image);
               }
            }, this)).catch(function (err) {
               console.log(err.stack);
            });
         }, this));
      }
   },

   /**
     * Cuts the sprites from the source image using information from a cgs and * cgg data and turns them into a strip
     * @param cgsPath - string
     * @param frames - array[array[CggPart]]
     * @param img - JIMP image?
     */
   makeStrip: function (cgsPath, frames, img) {
      var verbose = this.verbose;
      console.info('Loading ' + cgsPath);

      var columns = this.columns;
      var outputPath = this.outputPath;
      var includeEmpty = this.includeEmpty;

      var animation = path.parse(cgsPath).name;

      var ffbeScope = this;

      var json = {};

      return fs.readFileAsync(cgsPath, 'utf8')
         .then(function (data) {
            var topLeft = null;
            var bottomRight = null;
            var frameImages = [];
            var frameRect = null;

            var datasplit = data.replace('\r').split('\n');

            json.delay = [];

            var processDataLine = function (line, index) {
               return new Promise(function (resolve, reject) {
                  var params = line.split(',');
                  params.splice(params.length-1, 1);

                  var frameIndex, xPos, yPos, delay, frameImage;

                  if (params.length < 2) {
                     if (verbose) {
                        console.log('params.length was less than 2');
                     }

                     // resolving this as null, otherwise the entire
                     // Promise is rejected
                     return resolve(null);
                  }

                  frameIndex = parseInt(params[0]);
                  xPos = parseInt(params[1]);
                  yPos = parseInt(params[2]);
                  delay = parseInt(params[3]);

                  json.delay.push(delay);

                  createImage(2000, 2000).then(function (image) {
                     _.each(frames[frameIndex], function (part, idx) {
                        var crop;
                        var clone = img.clone(); // NOTE: crop is destructive, so we must reclone
                        crop = clone.crop(part.imgX, part.imgY, part.imgWidth, part.imgHeight);

                        if (part.blendMode === 1) {
                           if (verbose) console.log(' -- blending part -- ' );
                           crop = blend(crop);
                        }

                        if (part.flipX || part.flipY) {
                           if (verbose) console.log(' -- flipping horizontal: ' + part.flipX + ', vertical: ' + part.flipY);
                           crop.flip(part.flipX, part.flipY);
                        }

                        if (part.rotate !== 0) {
                           if (verbose) console.log(' -- rotating part: ' + part.rotate);
                           // NOTE: Jimp rotates clockwise whereas the cgg setting for
                           // rotation is given in terms of counter-clockwise rotation
                           // So multiply by -1 to reverse it.
                           crop.rotate(-1 * part.rotate, true);
                        }

                        if (part.opacity < 100) {
                           if (verbose) console.log(' -- reducing opacity: ' + part.opacity);
                           crop.opacity(part.opacity / 100);
                        }

                        if (verbose) {
                           console.log(' -- writing part ' + idx + ' of Frame ' + frameIndex + ' from line ' + index);
                        } 
                        image.composite(crop, 2000/2 + part.xPos + xPos, 2000/2 + part.yPos + yPos);
                     }); // end part.each

                     var rect = getColorBoundsRect(image, 0xFF000000, 0, false);
                     var frameObject = {
                        image: image,
                        rect: rect
                     };

                     if (rect.width > 0 && rect.height > 0) {
                        frameImages.push(frameObject);

                        if (topLeft === null) {
                           topLeft = {x: rect.x, y: rect.y};
                           bottomRight = {
                              x: rect.x + rect.width,
                              y: rect.y + rect.height
                           };
                        } else {
                           topLeft.x = Math.min(rect.x, topLeft.x);
                           topLeft.y = Math.min(rect.y, topLeft.y);
                           bottomRight.x = Math.max(rect.x + rect.width, bottomRight.x);
                           bottomRight.y = Math.max(rect.y + rect.height, bottomRight.y);
                        }

                        console.log('Frame ' + frameImages.length + ' of ' + animation + ' done');
                     } else if (includeEmpty) {
                        frameImages.push(frameObject);
                     } // end if rect.width > 0 and rect.height > 0


                     resolve(image);
                  }); // end createImage.then
               }); // end Promise
            }; // end processDataLine

            var processing = datasplit.map(processDataLine)
            var results = Promise.all(processing);

            return results.then(function (image) {
               console.info('--- Making ' + animation + ' strip ---');

               frameRect = {
                  x: topLeft.x - 5,
                  y: topLeft.y - 5,
                  width: bottomRight.x - topLeft.x + 10,
                  height: bottomRight.y - topLeft.y + 10
               };

               json.frameDimensions = frameRect;

               var animImage = null;
               var tmpColumns = columns;
               var rows = Math.ceil(frameImages.length / columns);

               if (columns === 0 || columns >= frameImages.length) {
                  columns = frameImages.length;
                  json.imageWidth = frameImages.length * frameRect.width;
                  json.imageHeight = frameRect.height;

                  createImage(frameImages.length * frameRect.width, frameRect.height)
                     .then(function (image) {
                        _.each(_.range(frameImages.length), function (index) {
                           var frameObject = frameImages[index];
                           var frame = frameObject.image;
                           var rect = frameObject.rect;
                           frame.crop(frameRect.x, frameRect.y, frameRect.width, frameRect.height);

                           if (verbose) {
                              console.log('compositing frame ' + (index + 1) + ' to strip');
                           }

                           image.composite(frame, index * frameRect.width, 0);
                        }); // end each frame

                        return image;
                     }) // end createImage.then
                     .then(function (image) {
                        var retObject = {
                           outputPath: outputPath,
                           cgsPath: cgsPath,
                           image: image,
                           json: json
                        };

                        if (outputPath !== '.') {
                           return mkdirp.mkdirpAsync(outputPath).then(function (directory) {
                              return Promise.resolve(retObject);
                           });
                        } else {
                           return Promise.resolve(retObject);
                        }
                     })
                     .then(_.bind(ffbeScope.saveFile, ffbeScope));
               } else {
                  json.imageWidth = columns * frameRect.width;
                  json.imageHeight = rows * frameRect.height;

                  createImage(columns * frameRect.width, rows * frameRect.height)
                     .then(function (image) {
                        _.each(_.range(rows), function (row) {
                           _.each(_.range(columns), function (col) {
                              var index = (row * columns) + col;
                              var frameObject = frameImages[index];

                              var frame, rect;

                              if (frameObject) {
                                 frame = frameObject.image;
                                 rect = frameObject.rect;

                                 frame.crop(frameRect.x, frameRect.y, frameRect.width, frameRect.height);

                                 if (verbose) {
                                    console.log('compositing frame ' + (index + 1) + ' to sheet');
                                 }
                                 image.composite(frame, col * frameRect.width, row * frameRect.height);
                              }
                           }); // end each col
                        }); // end each row

                        return image;
                     }) // end createImage.then
                     .then(function (image) {
                        var retObject = {
                           outputPath: outputPath,
                           cgsPath: cgsPath,
                           image: image,
                           json: json
                        };

                        if (outputPath !== '.') {
                           return mkdirp.mkdirpAsync(outputPath).then(function (directory) {
                              return Promise.resolve(retObject);
                           });
                        } else {
                           return Promise.resolve(retObject);
                        }
                     })
                     .then(_.bind(ffbeScope.saveFile, ffbeScope));
               } // end if-else
            }); // end results.then

         }); // end readFileAsync
   }, // end makeStrip

   saveFile: function (saveObject) {
      var pathObject = path.parse(saveObject.cgsPath);
      var bits = pathObject.name.split('_cgs_');
      var action = bits[0].substring('unit_'.length);
      var uid = bits[1];

      var filename, outputName;

      if (this.customFilename) {
         filename = uid + '_' + action;
      } else {
         filename = 'unit_' + action + '_' + uid;
      }
      outputName = path.join(saveObject.outputPath, filename + '.png');

      console.log('saving sprite strip : ' + outputName);
      saveObject.image.write(outputName);

      if (this.saveJson) {
         saveObject.json.animName = action;
         saveObject.json.unitId = uid;

         outputName = path.join(saveObject.outputPath, filename + '.json');
         console.log('saving json: ' + outputName);
         fs.writeFileAsync(outputName, JSON.stringify(saveObject.json))
            .catch(function (err) {
               console.error('Error trying to write json file: ' + filename);
               console.log(err);
            });
      }
   }
};

var main = function (argv) {
   var ffbe = new ffbeTool();

   if (argv.length < 3) {
      console.info(usage);
      return;
   }

   ffbe.id = parseInt(argv[2]);
   if (isNaN(ffbe.id)) {
      console.info(ffbe.usage);
      return;
   }

   ffbe.processCommandArgs(process.argv);

   ffbe.readCggAsync(ffbe.id)
      .then(_.bind(ffbe.readPngAsync, ffbe))
      .catch(function (err) {
         console.error(err);
      });
};

main(process.argv);
